### PR TITLE
ApiClient QA

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary
After we got serveral issues about NREs which were often caused by missing clients i did another full review on the ApiClient

# Details
Adds serveral missing clients, use PrepareMessage everytime suitable, add groupDm to the dm cache

# Notes
Things i had questions about:
- [x]  2 Methods return TransportUser instead of DiscordUser:
   - [x]  `internal async ValueTask<IReadOnlyList<TransportMember>> ListGuildMembersAsync`
   - [x] `internal async ValueTask<TransportUser> ModifyCurrentUserAsync`
 - [x] Line 1833: There is a check for threadChannels. Shouldnt this be handled by the custom jsonserializer?
 - [x] Line 2707: When we create a DmChannel we add it to the client but we dont do it for groupDms. Is this correct?
 - [x] Most of the time we use the method to prepare messages but there are some exceptions. Are those old and should be exchanged or is this intended?

### CODE REVIEW

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 1]
```diff
@@ -1824,13 +1827,7 @@ ulong channelId
...
         RestResponse res = await this._rest.ExecuteRequestAsync(request);
 
-        DiscordChannel ret = JsonConvert.DeserializeObject<DiscordChannel>(res.Response!)!;
-
-        // this is really weird, we should consider doing this better
-        if (ret.IsThread)
-        {
-            ret = JsonConvert.DeserializeObject<DiscordThreadChannel>(res.Response!)!;
-        }
+        DiscordChannel ret = DiscordJson.ToDiscordObject<DiscordThreadChannel>(res.Response!)!;
```
Comment (akiraveliara) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:1830]: tested?

**Review by akiraveliara — COMMENTED** [review]
if that one thing is tested and works fine feel free to merge

File: DSharpPlus/Net/Rest/DiscordApiClient.cs (hunk 1 above)
Comment (Plerx2493) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs:1830]: will test when the polly issue is fixed

**Comment by Plerx2493** [comment]
> Ok now there is a issue parsing DiscordWebhooks.....

```log
Unhandled exception. Newtonsoft.Json.JsonSerializationException: Error converting value "{"application_id":null,"avatar":null,"channel_id":"1041829106427367495","guild_id":"938120155974750288","id":"1041829141055541248","name":"Captain Hook","type":1,"user":{"id":"350967844957192192","username":"plerx0715","a
vatar":"25b462f92b6eecc736ec5fb5d2505755","discriminator":"0","public_flags":4194432,"flags":4194432,"banner":null,"accent_color":667654,"global_name":"Plerx","avatar_decoration_data":null,"banner_color":"#0a3006"},"token":"REDACTED","url":"https://d
iscord.com/api/webhooks/1041829141055541248/TOKEN-REDACTED"}
" to type 'DSharpPlus.Entities.DiscordWebhook'. Path ''.
 ---> System.ArgumentException: Could not cast or convert from System.String to DSharpPlus.Entities.DiscordWebhook.
   at Newtonsoft.Json.Utilities.ConvertUtils.EnsureTypeAssignable(Object value, Type initialType, Type targetType)
   at Newtonsoft.Json.Utilities.ConvertUtils.ConvertOrCast(Object initialValue, CultureInfo culture, Type targetType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
   --- End of inner exception stack trace ---
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
   at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType, JsonSerializer jsonSerializer)
   at Newtonsoft.Json.Linq.JToken.ToObject[T](JsonSerializer jsonSerializer)
   at DSharpPlus.Net.Serialization.DiscordJson.ToDiscordObject[T](JToken token)
   at DSharpPlus.Net.DiscordApiClient.GetWebhookWithTokenAsync(UInt64 webhookId, String webhookToken)
   at DSharpPlus.DiscordWebhookClient.AddWebhookAsync(UInt64 id, String token)
   ```

File: DSharpPlus/Net/Rest/DiscordApiClient.cs (hunk 1 above)
Comment (Plerx2493) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs]: tested and works flawless (tested all types)

**Review by akiraveliara — APPROVED** [review]

**Pull request merged**